man/man1/tau2slog2.1
.\" ** You probably do not want to edit this file directly **
.\" It was generated using the DocBook XSL Stylesheets (version 1.69.1).
.\" Instead of manually editing it, you probably should edit the DocBook XML
.\" source for it and then use the DocBook XSL Stylesheets to regenerate it.
.TH "TAU2SLOG2" "1" "04/25/2011" "" "Tools"
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.SH "NAME"
tau2slog2 \- convert TAU tracefiles to SLOG2 tracefiles
.SH "SYNOPSIS"
.HP 10
\fBtau2slog2\fR [\fBoptions\fR] {\fItau_tracefile\fR} {\fItau_eventfile\fR} {\fB\-o\fR\ \fIoutput.slog2\fR}
.SH "DESCRIPTION"
.PP
This program is generated when TAU is configured with the \-slog2 or \-slog2=<slog2_dir> option.
.PP
The tau2slog2 trace converter takes a single tau trace file (*.trc) and event definition file (*.edf) and produces a corresponding slog2 binary trace file (*.slog2).
.PP
The tau2slog2 converter is called from the command line with the locations of the tau trace and event files. These arguments must be followed by the \-o flag and the name of the slog2 file to be written. tau2slog 2 accepts no other arguments.
.SH "OPTIONS"
.PP
\fB[\-h|\-\-h|\-help|\-\-help]\fR
Display HELP message.
.PP
\fB [\-tc]\fR
Check increasing endtime order, exit when 1st violation occurs.
.PP
\fB [\-tcc]\fR
Check increasing endtime order,continue when violations occur.
.PP
\fB [\-nc number]\fR
Number of childern per node (default is 2)
.PP
\fB [\-ls number]\fR
Max byte size of leaf nodes (default is 65536)
.PP
\fB [\-o output.slog2]\fR
Output filename with slog2 suffix
.SH "NOTES"
.PP
The max byte size of leaf nodes can be specified with suffix k, K, m or M, where k or K stands for kilobyte, m or M stands for megabyte. e.g. 64k means 65536 bytes.
.SH "EXAMPLES"
.PP
A typical invocation of the converter, to create app.slog2, is as follows:
.sp
.nf
tau2slog2 app.trc tau.edf \-o app.slog2
      
.fi
.sp
.SH "SEE ALSO"
.PP
tau_merge(1),
tau_convert(1)


Changes
*****************************************************************************
**                      TAU Portable Profiling Package                     **
**                      http://tau.uoregon.edu                             **
*****************************************************************************
**    Copyright 1997-2011						   **
**    Department of Computer and Information Science, University of Oregon **
**    Research Centre Juelich, Germany                                     **
**    Advanced Computing Laboratory, Los Alamos National Laboratory        **
*****************************************************************************

Change log:
------------
Version 2.20 changes (from 2.20):
1. Added support for GPGPUs using tau_exec -cuda and tau_exec -opencl. Configure with -cuda=<dir>.
2. Added support for 3D topology displays in paraprof. 
3. Added support for tracking per-communicator performance data -PROFILECOMMUNICATORS.
4. Added support for binary rewriting of static executables using DyninstAPI 7.0. 
5. Added support for derived metrics, cut, copy and paste of metrics in paraprof, perfexplorer.
6. Added support for sampling based measurements in tau_exec -ebs* (README.sampling).
7. Added support for ARMCI profiling (GA v5.0) in tau_exec -armci. Configure with -armci=<dir>.
8. Added support for loop level instrumentation in the binary rewriter (tau_run a.out -o a.i -f select.tau).
9. Added preliminary support for Eclipse PTP remote components. 
10. Integrated tau_exec and tau_wrap (tau_exec -loadlib=<dir>/libwrapped.so) for instrumenting an external library.
11. Added support for pre-computing mean and std. dev. at the end of execution with TAU_PROFILE_FORMAT=merged.
12. Added support for compiler-based instrumentation for Cray CCE compilers. 
13. TAU ported to Cray XE6 with updated default directories (gemini).  
14. Improved support for Cray Shmem. 
15. Improved support for thread based communication in tau2otf and ParaVer support in tau_convert -paraver.
16. Throttled functions are now marked explicitly in their names [THROTTLED]. 
17. Before reverting to compiler based instrumentation the user is prompted unless -optRevert is specified. 
18. Updates to memory and I/O tracking support for Mac OS X (in tau_exec).
19. Added support for the Yorick programming language. 
20. Added Java profiling support using JVMTI.
21. Added support for Score-P (www.score-p.org) measurement system. Configure with -scorep=<dir>.
22. Added support for importing Google perftools performance data (paraprof -f google).
23. Added support for profiling CUDA kernels on a separate thread.
24. Bug fixes (papi+tracing, profile merged with tau_exec -io, outer-loop level instrumentation dyinstAPI)
25. Added support for topology display with user specified topologies loadable from a text file in paraprof. 
26. Added support for PGI 11.x compilers with support for accelerator primitives. 
27. Added support for a scrollbar in the paraprof 3D window.
28. CUDA 4.0 support added and tracking of gpgpu thread execution is now supported. 
29. Bug fixes for PGI -optCompInst for C++, tau_exec -memory for Intel -optCompInst. 
30. PostgreSQL jar file support updated in PerfDMF. 
31. Added support for Score-P LD_PRELOAD'ing using tau_exec and tau_run.
32. Added support for SOL2CC compiler for Solaris CC compiler. 
33. Bug fixes for tau_instrumentor (single line Fortran DO loop instrumentation).
34. Added support for Cray CCE -optCompInst for OpenMP.
35. Bug fix for PAPI and TAU_TRACE=1 for initialization of papi. 
36. Added support for demangling kernel names for CUDA executions. 

-------------------------------
Version 2.19 changes (from 2.18):
1. Added support for Chapel.
2. Added support for UPC (Berkeley upcc compiler, GASP, supports -optCompInst).
3. Added support for automatically generating LD_PRELOADable .so files from a .h file using tau_wrap (-r libname.so).
4. Derived metrics window in ParaProf and PerfExplorer.
5. -bfd=download now downloads binutils-2.20 instead of 2.18.
6. Added an aligned stacked bar chart in PerfExplorer that is similar to unchecking "Stack Bars Together" box in paraprof's options window.
7. Paraprof now automatically adjusts the memory used (tau_javamax.sh)
8. Added support for Intel compilers on Cray XT systems.
9. tau_validate now uses TAU_VALIDATE_PARALLEL and TAU_VALIDATE_SERIAL env vars to run the tests (see --help). 
10. Added support for external tool configuration in PerfExplorer.
11. Updated PerfExplorer code to Weka 3.6.1.
12. Added support for DBSCAN clustering.
13. Updated Jython support to 2.5.1 that supports Python v2.5. 
14. Created a utility to reconstruct a Paraver trace from TAU EBS samples.
15. Paraprof 3D communication matrix display has cross hairs and value boxes.
16. Enabled tree selection model for multi-selection.
17. New expression parsing window in Paraprof.
18. Paraprof 3D windows now work on IBM BG/P using ppc64 JOGL.
19. Group changer window in paraprof. 
20. Added support for outer loop level instrumentation in tau_instrumentor's spec file mode. 
21. When PDT based source instrumentation fails, compiler-based instrumentation is used as a fallback. Disable with -optNoCompInst in TAU_OPTIONS env. var.
22. Added support for PAPI-C v4.0 in TAU. Retains backward compatibility with earlier PAPI versions. 
23. Added support for Cray CCE compilers on XT systems (module PrgEnv-cray).
24. Added support for tracking pthread barrier wait times. 
25. Added support for TAU over MRNet.
26. Added a new tool (tau_exec) to evaluate I/O, memory and communication
27. PerfExplorer has a new derived metric pane and updates to configuration
28. tau_exec can also load wrapper libraries created by tau_wrap using -loadlib
29. Added support for sampling based profiling (README.sampling)
30. Added support for SHMEM wrappers for Cray XT
31. Added support for Cray XMT (-arch=crayxmt)
32. Added support for ScoreP (aka silc).

tools/src/tau2slog2/src/edu/uoregon/tau/Tau2Slog2.java
/*
 *  See TAU License file
 */

/*
 * @author  Wyatt Spear
 * Derived from code by Anthony Chan
 */

package edu.uoregon.tau;

//import java.util.;


import java.util.Date;
import java.util.HashMap;
import java.util.Map;

import base.drawable.*;
//import logformat.trace.*;
import logformat.slog2.*;
import logformat.slog2.output.TreeTrunk;

public class Tau2Slog2
{
    /*static {
        System.loadLibrary( "TraceInput" ); 
    }*/

    private static short    num_children_per_node = 0;
    private static int      leaf_bytesize         = 0;
    private static String   tautrc, tauedf, out_filename;
    private static boolean  enable_endtime_check;
    private static boolean  continue_when_violation;

    @SuppressWarnings("unchecked")
	public static void main( String[] args )
    {
        InputLog							dobj_ins;
        logformat.slog2.output.OutputLog    slog_outs;
        Kind                                next_kind;
        Topology                            topo;
        CategoryMap                         objdefs;   // Drawable def'n
        Map<Topology,Category>                                 shadefs;   // Shadow   def'n
        Category                            objdef;
        LineIDMapList                       lineIDmaps;
        LineIDMap                           lineIDmap;
        Primitive                           prime_obj;
        Composite                           cmplx_obj;
        long                                Nobjs;

        TreeTrunk                           treetrunk;
        double                              prev_dobj_endtime;
        double                              curr_dobj_endtime;
        long                                offended_Nobjs;
        Drawable                            offended_dobj;



        //  Initialize prev_dobj_endtime to avoid complaint by compiler
        prev_dobj_endtime = Double.NEGATIVE_INFINITY;
        offended_Nobjs    = Integer.MIN_VALUE;
        offended_dobj     = null;

        out_filename = null;
        parseCmdLineArgs( args );
        if ( out_filename == null ){
        	int lastP=tautrc.lastIndexOf('.');
        	String outF =tautrc.substring(0, lastP);
        	outF+=".slog2";
        	out_filename=outF;
        	 //out_filename  = TraceName.getDefaultSLOG2Name( tautrc );
        }
           

        objdefs       = new CategoryMap();
        shadefs       = new HashMap<Topology,Category>();
        lineIDmaps    = new LineIDMapList();
        Nobjs         = 0;

        /* */    Date time1 = new Date();
        dobj_ins   = new InputLog( tautrc,tauedf );
        slog_outs  = new logformat.slog2.output.OutputLog( out_filename );

        //  Set Tree properties, !optional!
        //  TreeNode's minimum size, without any drawable/shadow, is 38 bytes.
        //  Drawable;s minimum size is 32 bytes, whether it is state/arrow.
        //  Arrow( with 2 integer infovalues ) is 40 bytes long.
        //  So, for 1 state primitive leaf, the size is 38 + 40 = 78 .
        if ( leaf_bytesize > 0 )
            slog_outs.setTreeLeafByteSize( leaf_bytesize );
        if ( num_children_per_node > 0 )
            slog_outs.setNumChildrenPerNode( num_children_per_node );

        treetrunk = new TreeTrunk( slog_outs, shadefs );
        /* */    Date time2 = new Date();
        while ( ( next_kind = dobj_ins.peekNextKind() ) != Kind.EOF ) {
            if ( next_kind == Kind.TOPOLOGY ) {
                topo = dobj_ins.getNextTopology();
                objdef = Category.getShadowCategory( topo );
                objdefs.put( new Integer( objdef.getIndex() ), objdef );
                shadefs.put( topo, objdef );
            }
            else if ( next_kind == Kind.YCOORDMAP ) {
                lineIDmap = new LineIDMap( dobj_ins.getNextYCoordMap() );
                lineIDmaps.add( lineIDmap );
            }
            else if ( next_kind == Kind.CATEGORY ) {
                objdef = dobj_ins.getNextCategory();
                objdefs.put( new Integer(objdef.getIndex() ),objdef );
            }
            else if ( next_kind == Kind.PRIMITIVE ) {
                prime_obj = dobj_ins.getNextPrimitive();
                //System.out.println(objdefs);
                //System.out.println(prime_obj);
                prime_obj.resolveCategory( objdefs );
                // postponed, wait till on-demand decoding of InfoBuffer
                // prime_obj.decodeInfoBuffer();
                Nobjs++;
                // System.out.println( Nobjs + " : " + prime_obj );
                if ( enable_endtime_check ) {
                    if ( ! prime_obj.isTimeOrdered() ) {
                        System.out.println( "**** Primitive Time Error ****" );
                        if ( ! continue_when_violation )
                            System.exit( 1 );
                    }
                    curr_dobj_endtime = prime_obj.getLatestTime();
                    if ( prev_dobj_endtime > curr_dobj_endtime ) {
                        System.err.println( "**** Violation of "
                                          + "Increasing Endtime Order ****\n"
                                          + "\t Offended Drawable -> "
                                          + offended_Nobjs + " : "
                                          + offended_dobj + "\n"
                                          + "\t Offending Primitive -> "
                                          + Nobjs + " : " + prime_obj + "\n"
                                          + "   previous drawable endtime ( "
                                          + prev_dobj_endtime + " ) "
                                          + " > current drawable endtiime ( "
                                          + curr_dobj_endtime + " ) " );
                        if ( ! continue_when_violation )
                            System.exit( 1 );
                    }
                    offended_Nobjs    = Nobjs;
                    offended_dobj     = prime_obj;
                    prev_dobj_endtime = curr_dobj_endtime;
                }
                treetrunk.addDrawable( prime_obj );
            }
            else if ( next_kind == Kind.COMPOSITE ) {
                cmplx_obj = dobj_ins.getNextComposite();
                cmplx_obj.resolveCategory( objdefs );
                // postponed, wait till on-demand decoding of InfoBuffer
                // cmplx_obj.decodeInfoBuffer();
                Nobjs++;
                // System.out.println( Nobjs + " : " + cmplx_obj );
                if ( enable_endtime_check ) {
                    if ( ! cmplx_obj.isTimeOrdered() ) {
                        System.out.println( "**** Composite Time Error ****" );
                        if ( ! continue_when_violation )
                            System.exit( 1 );
                    }
                    curr_dobj_endtime = cmplx_obj.getLatestTime();
                    if ( prev_dobj_endtime > curr_dobj_endtime ) {
                        System.err.println( "***** Violation of "
                                          + "Increasing Endtime Order! *****\n"
                                          + "\t Offended Drawable -> "
                                          + offended_Nobjs + " : "
                                          + offended_dobj + "\n"
                                          + "\t Offending Composite -> "
                                          + Nobjs + " : " + cmplx_obj + "\n"
                                          + "   previous drawable endtime ( "
                                          + prev_dobj_endtime + " ) "
                                          + " > current drawable endtiime ( "
                                          + curr_dobj_endtime + " ) " );
                        if ( ! continue_when_violation )
                            System.exit( 1 );
                    }
                    offended_Nobjs    = Nobjs;
                    offended_dobj     = cmplx_obj;
                    prev_dobj_endtime = curr_dobj_endtime;
                }
                treetrunk.addDrawable( cmplx_obj );
            }
            else {
                System.err.println( "Tau2Slog2: Unrecognized return "
                                  + "from peekNextKind() = " + next_kind );
            }
        }   // Endof while ( dobj_ins.peekNextKind() )
        treetrunk.flushToFile();

        objdefs.removeUnusedCategories();
        slog_outs.writeCategoryMap( objdefs );

        lineIDmaps.add( treetrunk.getIdentityLineIDMap() );
        slog_outs.writeLineIDMapList( lineIDmaps );

        slog_outs.close();
        dobj_ins.close();

        /* */    Date time3 = new Date();
        System.out.println( "\n" );
        System.out.println( "Number of Drawables = " + Nobjs );

        // System.out.println( "time1 = " + time1 + ", " + time1.getTime() );
        // System.out.println( "time2 = " + time2 + ", " + time2.getTime() );
        // System.out.println( "time3 = " + time3 + ", " + time3.getTime() );
        System.out.println( "timeElapsed between 1 & 2 = "
                          + ( time2.getTime() - time1.getTime() ) + " msec" );
        System.out.println( "timeElapsed between 2 & 3 = "
                          + ( time3.getTime() - time2.getTime() ) + " msec" );
    }

    private static int parseByteSize( String size_str )
    {
        int idxOfKilo = Math.max( size_str.indexOf( 'k' ),
                                  size_str.indexOf( 'K' ) );
        int idxOfMega = Math.max( size_str.indexOf( 'm' ),
                                  size_str.indexOf( 'M' ) );
        if ( idxOfKilo > 0 )
            return Integer.parseInt( size_str.substring( 0, idxOfKilo ) )
                   * 1024;
        else if ( idxOfMega > 0 )
            return Integer.parseInt( size_str.substring( 0, idxOfMega ) )
                   * 1024 * 1024;
        else
            return Integer.parseInt( size_str );
    }

    private static String help_msg = "\nUsage: tau2slog "
                                   + "[options] <tau_trc_file> <tau_edf_file> -o <output.slog2>\n"
                                   + " options: \n"
                                   + "\t [-h|--h|-help|--help]             "
                                   + "\t Display HELP message.\n"
                                   + "\t [-tc]                             "
                                   + "\t Check increasing endtime order,\n"
                                   + "\t                                   "
                                   + "\t exit when 1st violation occurs.\n"
                                   + "\t [-tcc]                            "
                                   + "\t Check increasing endtime order,\n"
                                   + "\t                                   "
                                   + "\t continue when violations occur.\n"
                                   + "\t [-nc number_of_children_per_node] "
                                   + "\t Default value is "
                                   + logformat.slog2.Const.NUM_LEAFS +".\n"
                                   + "\t [-ls max_byte_size_of_leaf_node]  "
                                   + "\t Default value is "
                                   + logformat.slog2.Const.LEAF_BYTESIZE +".\n"
                                   + "\t [-o output_filename_with_slog2_suffix]"
                                   + "\n\n"
                                   + " note: \"max_byte_size_of_leaf_node\" "
                                   + "can be specified with suffix "
                                   + "k, K, m or M,\n"
                                   + "       where k or K stands for kilobyte,"
                                   + " m or M stands for megabyte.\n"
                                   + "       e.g. 64k means 65536 bytes.\n"
    + "\n"
    + "Example:\n"
    + "  tau2slog2 tau.trc tau.edf -o tau.slog2\n\n";

    private static void parseCmdLineArgs( String argv[] )
    {
        String        arg_str;
        int           idx;
        StringBuffer  err_msg = new StringBuffer();

        tautrc = null;
        tauedf = null;
        enable_endtime_check     = false;
        continue_when_violation  = false;

        if ( argv.length == 0 ) {
            System.out.println( help_msg );
            System.out.flush();
            System.exit( 0 );
        }

        idx = 0;
        try {
            while ( idx < argv.length ) {
                if ( argv[ idx ].startsWith( "-" ) ) {
                    if (  argv[ idx ].equals( "-h" ) 
                       || argv[ idx ].equals( "--h" )
                       || argv[ idx ].equals( "-help" )
                       || argv[ idx ].equals( "--help" ) ) {
                        System.out.println( help_msg );
                        System.out.flush();
                        System.exit( 0 );
                    }
                    else if ( argv[ idx ].equals( "-tc" ) ) {
                        enable_endtime_check     = true;
                        continue_when_violation  = false;
                        err_msg.append( "\n endtime_order_check_exit = true" );
                        idx++;
                    }
                    else if ( argv[ idx ].equals( "-tcc" ) ) {
                        enable_endtime_check     = true;
                        continue_when_violation  = true;
                        err_msg.append( "\n endtime_order_check_stay = true" );
                        idx++;
                    }
                    else if ( argv[ idx ].equals( "-nc" ) ) {
                        arg_str = argv[ ++idx ]; 
                        num_children_per_node = Short.parseShort( arg_str );
                        err_msg.append( "\n number_of_children_per_node = "
                                      + arg_str );
                        idx++;
                    }
                    else if ( argv[ idx ].equals( "-ls" ) ) {
                        arg_str = argv[ ++idx ];
                        leaf_bytesize = parseByteSize( arg_str );
                        err_msg.append( "\n max_byte_size_of_leaf_node = "
                                      + arg_str );
                        idx++;
                    }
                    else if ( argv[ idx ].equals( "-o" ) ) {
                        out_filename = argv[ ++idx ].trim();
                        err_msg.append( "\n output_filename = "
                                      + out_filename );
                        idx++;
                        if ( ! out_filename.endsWith( ".slog2" ) )
                            System.err.println( "Warning: The suffix of the "
                                              + "output filename is NOT "
                                              + "\".slog2\"." );
                    }
                    else {
                        System.err.println( "Unrecognized option, "
                                          + argv[ idx ] + ", at "
                                          + indexOrderStr( idx+1 )
                                          + " command line argument" );
                        System.out.flush();
                        System.exit( 1 );
                    }
                }
                else {
                    tautrc   = argv[ idx ];
                    idx++;
                    tauedf = argv[idx];
                    idx++;
                }
            }
        } catch ( ArrayIndexOutOfBoundsException idxerr ) {
            if ( err_msg.length() > 0 )
                System.err.println( err_msg.toString() );
            System.err.println( "Error occurs after option "
                              + argv[ idx-1 ] + ", " + indexOrderStr( idx )
                              + " command line argument." );
            // System.err.println( help_msg );
            idxerr.printStackTrace();
        } catch ( NumberFormatException numerr ) {
            if ( err_msg.length() > 0 )
                System.err.println( err_msg.toString() );
           // String idx_order_str = indexOrderStr( idx );
            System.err.println( "Error occurs after option "
                              + argv[ idx-1 ] + ", " + indexOrderStr( idx )
                              + " command line argument.  It needs a number." );
            // System.err.println( help_msg );
            numerr.printStackTrace();
        }
        
        if ( tautrc == null || tauedf == null ) {
            System.err.println( "The Program needs a TAU trc and edf filename as "
                              + "command line arguments." );
            System.err.println( help_msg );
            System.exit( 1 );
        }
    }

    private static String indexOrderStr( int idx )
    {
        switch (idx) {
            case 1  : return Integer.toString( idx ) + "st";
            case 2  : return Integer.toString( idx ) + "nd";
            case 3  : return Integer.toString( idx ) + "rd";
            default : return Integer.toString( idx ) + "th";
        }
    }
}
